transports/tcp/src/lib.rs

```diff
@@ -734,116 +714,84 @@ where
-                                        log::debug!("New listen address: {}", ma);
-                                        me.port_reuse.register(ip, me.listen_addr.port());
-                                        return Poll::Ready(Some(Ok(
-                                            TcpListenerEvent::NewAddress(ma),
-                                        )));
-                                    }
-                                }
-                                Ok(IfEvent::Down(inet)) => {
-                                    let ip = inet.addr();
-                                    if me.listen_addr.is_ipv4() == ip.is_ipv4() && addrs.remove(&ip)
-                                    {
-                                        let ma = ip_to_multiaddr(ip, me.listen_addr.port());
-                                        log::debug!("Expired listen address: {}", ma);
-                                        me.port_reuse.unregister(ip, me.listen_addr.port());
-                                        return Poll::Ready(Some(Ok(
-                                            TcpListenerEvent::AddressExpired(ma),
-                                        )));
-                                    }
-                                }
-                                Err(err) => {
-                                    log::debug! {
-                                        "Failure polling interfaces: {:?}. Scheduling retry.",
-                                        err
-                                    };
-                                    me.pause = Some(Delay::new(me.sleep_on_error));
-                                    return Poll::Ready(Some(Ok(TcpListenerEvent::Error(err))));
-                                }
-                            }
-                        }
-                    }
-                },
-                // If the listener is bound to a single interface, make sure the
-                // address is registered for port reuse and reported once.
-                InAddr::One { addr, out } => {
-                    if let Some(multiaddr) = out.take() {
-                        me.port_reuse.register(*addr, me.listen_addr.port());
-                        return Poll::Ready(Some(Ok(TcpListenerEvent::NewAddress(multiaddr))));
-                    }
+        if let Some(mut pause) = me.pause.take() {
+            match pause.poll_unpin(cx) {
+                Poll::Ready(_) => {}
+                Poll::Pending => {
+                    me.pause = Some(pause);
+                    return Poll::Pending;
                 }
             }
+        }
 
-            if let Some(mut pause) = me.pause.take() {
-                match Pin::new(&mut pause).poll(cx) {
-                    Poll::Ready(_) => {}
-                    Poll::Pending => {
-                        me.pause = Some(pause);
-                        return Poll::Pending;
+        if let Some(if_watcher) = me.if_watcher.as_mut() {
+            loop {
```

Comment:
In our guidelines we usually put the loop around everything, not just one single child. Would that make sense here?

Reply:
I don't think it makes sense here. Apart from the `IfWatcher`, all other sections directly return. There is never a case where a child makes progress and then we want to `continue` from the top. 
We only really need the loop in the first place because the `IfWatcher` may return Ipv6 address events that we want to skip if we listen on an ipv4 address and the other way around. In that case we ignore the event and poll the `IfWatcher` again.  
Actually, now that I think about it again I think I even prefer the previous `while let Poll::Ready(event) = if_watcher.poll_if_event(cx)`. 

I think the whole loop pattern makes sense in implementations where we drive multiple futures that are interdependent, i.e. if one child makes progress and we handle its events, it could be that now previously polled children can make progress again. But this is not the case here.

Reply:
Fair enough. The only argument against would be that consistency makes it easier to read / skim the code, even if we would be able to write it in a different style here.

I don't mind much though :)